Droidcon-Boston/app/src/main/java/com/mentalmachines/droidcon_boston/firebase/AuthController.kt

```diff
@@ -39,15 +50,27 @@ class AuthController(@DrawableRes private val loginScreenAppIcon: Int) {
 
         return if (resultCode == Activity.RESULT_OK) {
             user = FirebaseAuth.getInstance().currentUser
+            fetchDataFromFirebase()
             null
         } else {
             response?.error?.message
         }
     }
 
     fun logout(context: Context) {
-        AuthUI.getInstance()
-            .signOut(context)
+        AuthUI.getInstance().signOut(context)
         user = null
     }
+
+    override fun onDataChange(dataSnapshot: DataSnapshot) {
+        if (user == null) {
+            //TODO clear the user cached in the Firebase database.
+        } else {
+            //TODO store an instance of the user in the Firebase database for use throughout app.
```

Comment:
@kenyee  I'm actually having a bit of trouble with the Firebase stuff, its not letting me authenticate with my gmail (ccorrads@gmail.com)... I confirmed I'm signing with the debug keystore with the same SHA1 I provided you...not sure. I'm not sure what the result looks like in this instance, so I left a few TODOs here.

Reply:
@CCorrado is there an error code when it fails?  Code 10 Error 10 means the SHA1 is wrong

Reply:
I also think you need per-user space in firebase to put that data.
It needs a node and permissions set on that node for you to dump the data into:
https://firebase.google.com/docs/database/security/user-security

Reply:
FYI, this is in the firebase permissions ruleset I added last year:

    "users": {
      // private data area for users
      "$uid": {
        ".read": "$uid === auth.uid",
        ".write": "$uid === auth.uid"
      }
    },
There's also a top level /users/ node in the DB

Reply:
I'm getting `com.firebase.ui.auth.FirebaseUiException: Code: 12500, message: 12500:`

I confirmed on my macbook pro Friday..... but I'm on my mac mini now. :) All is resolved and I'm able to log in w/ my gmail. User Error.

Reply:
Otherwise, looking at the result now, I think we have enough here to manage to add users to the firebase data store without having to store the whole auth object -- we only need (at most) from Firebase auth 
- email
- name
- photo

So far this is still a stepping stone to the Eventbrite piece, where we can support fetching the data that the user has filled out when they bought their ticket. (i.e. twitter handle, attendee status, etc)

Reply:
This is an important stepping stone to all the other stuff we want to add (favs, rating a session, etc. too).
That's why I prepped all this stuff in the Firebase permissions and DB already to make sure it was all ready for adding chat to the app last year.
What was the mistake you made? Would help knowing in case someone else hits the same thing; even if it sounds stupid, I've found that warning people helps...

What I'm not quite sure about yet is how we can read all the user's session ratings yet...seems like there'd need to be some sort of admin group or some periodic Cloud Function that runs through the user's private data area to summarize the session ratings.

Reply:
It turned out to be the wrong SHA1. My mac mini wasn't registered, just my mbp (the one i gave you)

Agreed. Well with this PR we can at least have a reference to user data via their registered e-mail. So if we want to store ratings, favorites, etc we can just stuff it into the User object data model. Was there something else you had in mind? Using the Firebase user object itself could be another way to do it maybe, but i haven't played with it .

Reply:
Wrong SHA would do it :-)  Odd that you didn't get the Error Code 10 I did when I had that issue.

Is the User object data model stored client side or server side?
With KotlinConf, they gave us a summary of reviews of the session.  We probably need to do something like that too.  In the database itself, there's a /users node and the permissions I mentioned above gives everyone access to a /users/<uid> node that you could dump info into.

Reply:
It would be synced with the Firebase database, so both!

Nice, I understand. As a user, I would want to be the only one with access to my user object. So there might be some permission rules we'd have to set up, or is that what this code does?

```
"$uid": {
    ".read": "$uid === auth.uid",
    ".write": "$uid === auth.uid"
}
```

Reply:
Yep..exactly that those permission do :-)
There's also the other node I mentioned...
"/users/$uid"...

Reply:
updated. Should be storing our User.kt data model in the /users/$uid container now.

Reply:
hmm...I don't see anything in that node

Reply:
sorry, i logged myself out lol. My profile is there now :)

Reply:
Now I see it..cool :-)